src/test.js

```diff
@@ -2,50 +2,38 @@ import assert from "assert";
 
 export default addTest;
 
-process.nextTick(runTests); // Wait until test file is parsed
-
 const suite = [];
-
 function addTest(title, implementation) {
-  const test = createTest(title, implementation);
+  const test = { title, implementation, error: null };
   suite.push(test);
 }
 
-function createTest(title, implementation) {
-  return {
-    title,
-    error: null,
-    result: undefined,
-    start() {
-      this.result = this.run();
-    },
-    async run() {
-      try {
-        await implementation({ // Pass in a test execution contest (t)
-          assert: assert.ok,
-          deepEqual: assert.deepStrictEqual,
-          is: assert.strictEqual
-        });
-      } catch (e) {
-        this.error = e;
-      }
-    }
-  };
-}
+process.nextTick(runTests); // Wait until test file is parsed
 
 async function runTests() {
   console.log("\nRunning...");
   console.time("Ran in");
-  for (const test of suite) {
-    test.start();
-  }
-  for (const test of suite) {
-    await test.result;
-    report(test);
-  }
+  await suite.map(startTest).reduce(reportSequentially, Promise.resolve());
```

Comment:
I would change `reportSequentially` to take the list of "started tests" and then use `for of` + `await`, for whatever reason I've always found that to be more obvious as a "run promises in sequence" rather than reducing over a list of promises.

```js
async function reportSequentially (runningTests) {
  for (const run of runningTests) {
   await run;
  }
}
```

This line would then become `await reportSequentially(suite.map(startTest))`

Reply:
Yeah, you're right, technically you're reducing an array of promises to a single promise so it's not too far from it's intended use, but `reduce` is typically not very straightforward.

I changed it in 19a0d80e47f714549e87ed5d17bd550b090d6674

The only con I see to `await reportSequentially(suite.map(startTest))` is that it you're reading the second action before the first.

I *really* wish I could do this:
```
suite |> startTests |> reportTestsSequentially
```
But the pipeline operator is still a stage 1 proposal  :(

Reply:
It depends on what ordering you want to be fair.

`reportTestsSequentially` now reports the tests in the order in which they were detected/started right?

Reply:
> `reportTestsSequentially` now reports the tests in the order in which they were detected/started right?

Yes, I just meant that if you had, say, something like `report(start(prepare(check(tests))))`  you, as a human, would be reading the order of operations backwards, last action is on the left, first on the right, whereas the pipe operator `tests |> check |> prepare |> start |> report` or a fluent interface like `tests.check().prepare().start().report()` read more naturally.

In the case of `reportSequentially(suite.map(startTest))` it's fine since it's just one level deep but it doesn't look as nice :( 

In any case, as I said, I went with your suggestion and pushed it here 19a0d80e47f714549e87ed5d17bd550b090d6674

Reply:
Yeah I saw that

I mean you could also use [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled)
```js
Promise.allSettled(suite.map(startTest)).then(testOutcome => {
  // do more stuff like reporting
})
```

But then you're waiting for all the tests to pass/fail before reporting...

Reply:
That's the thing, the idea is to report a test as soon as it's ready, but in the same order they were declared (so, basically a stream of results). 

Some libraries have things like `Promise.each` or `async.eachSeries`, but for vanilla JS it's either a `for` loop or `reduce` :/